lib/ansible/playbook/role/__init__.py

```diff
@@ -530,18 +546,21 @@ def get_vars(self, dep_chain=None, include_params=True, only_exports=False):
     def get_direct_dependencies(self):
         return self._dependencies[:]
 
+    def __hash__(self):
+        return hash((self._role_collection, self._role_path, self._role_name))
+
     def get_all_dependencies(self):
         """
         Returns a list of all deps, built recursively from all child dependencies,
         in the proper order in which they should be executed or evaluated.
         """
         if self._all_dependencies is None:
 
-            self._all_dependencies = []
+            self._all_dependencies = set()
```

Comment:
while sets are more efficient, they are also 'unordered' so this will break expected variable precedence  and will change the display of 'role name' variables

Reply:
I changed the implementation to use a list again, but check whether the dep is already in it. Would this be alright regarding var precedence? It seemed to work ok in my test run at least.

Reply:
> add dupe, but remove previous instance

This sounds to fit the previous behavior, where vars that occurred multiple times overwrite themselves until the list is fully iterated over.

Reply:
I changed it to remove previous instances. Guess this is reflecting the current behavior in the correct way. Having a test would surely help here.

Reply:
I extended the `role_dep_chain` test to include different types of variables, including overwriting them in parent roles. Please check 35e35eb928629086d7c85e829ede1d3b00c3539e and let me know if this fits.

Reply:
Made a minor copy/paste error when pasting the test into my machine connected to GitHub. Should be fixed now, let's see whether the tests now pass.